File: altstu/altstu/athleteaccount.cpp
//
//  athleteaccount.cpp
//  altstu
//
//  Created by Stepan Bondarenko on 04.11.2025.
//

#include "athleteaccount.hpp"
#include <iostream>

// Инициализация статического поля
int AthleteAccount::s_accountCount = 0;

AthleteAccount::AthleteAccount(const std::string& accountId, const std::string& displayName, GoalCode primaryGoal)
    : m_accountId(accountId), m_displayName(displayName), m_primaryGoal(primaryGoal), m_experienceBand(0)
{
    // Использование this для доступа к полям
    this->m_accountId = accountId;
    this->m_displayName = displayName;
    s_accountCount++; // Увеличение статического счетчика
}

AthleteAccount::AthleteAccount(const AthleteAccount& other)
    : m_accountId(other.m_accountId),
      m_displayName(other.m_displayName),
      m_primaryGoal(other.m_primaryGoal),
      m_experienceBand(other.m_experienceBand),
      m_timeWindows(other.m_timeWindows),
      m_dietStyleCode(other.m_dietStyleCode),
      m_allowedGear(other.m_allowedGear)
{
    s_accountCount++; // Увеличение статического счетчика при копировании
}

AthleteAccount::~AthleteAccount()
{
    s_accountCount--; // Уменьшение статического счетчика
}

void AthleteAccount::updateGoal(GoalCode newGoal)
{
    GoalCode oldGoal = m_primaryGoal;
    m_primaryGoal = newGoal;
    logGoalChange(oldGoal, newGoal);
}

void AthleteAccount::setTimeWindows(const TimeWindowsType& windows)
{
    m_timeWindows = windows;
}

bool AthleteAccount::prefers(const std::string& gear) const
{
    return m_allowedGear.find(gear) != m_allowedGear.end();
}

void AthleteAccount::logGoalChange(GoalCode oldGoal, GoalCode newGoal)
{
    std::cout << TAG_FOR_LOG << ": Goal changed from " << static_cast<int>(oldGoal) 
              << " to " << static_cast<int>(newGoal) << std::endl;
}

File: altstu/AltstuLab4/Domain/AthleteAccount.cs
using System;
using System.Collections.Generic;

namespace AltstuLab4.Domain
{
    public sealed class AthleteAccount
    {
        public const int MaxExperienceBand = 2;

        // Статическое поле - счетчик созданных аккаунтов
        private static int _accountCount = 0;

        // Статический метод для получения количества аккаунтов
        public static int GetAccountCount() => _accountCount;

        private readonly List<TimeWindow> _timeWindows = new();
        private readonly HashSet<string> _allowedGear = new(StringComparer.OrdinalIgnoreCase);

        public AthleteAccount(string accountId, string displayName, GoalCode primaryGoal)
        {
            // Использование this для явного указания на поля класса
            this.AccountId = accountId ?? throw new ArgumentNullException(nameof(accountId));
            this.DisplayName = displayName ?? throw new ArgumentNullException(nameof(displayName));
            this.PrimaryGoal = primaryGoal;
            
            // Увеличение статического счетчика
            _accountCount++;
        }

        public string AccountId { get; }
        public string DisplayName { get; }
        public GoalCode PrimaryGoal { get; private set; }
        
        // Свойство с валидацией в set
        private int _experienceBand;
        public int ExperienceBand
        {
            get => _experienceBand;
            set
            {
                if (value < 0 || value > MaxExperienceBand)
                {
                    throw new ArgumentOutOfRangeException(nameof(value), 
                        $"ExperienceBand must be between 0 and {MaxExperienceBand}");
                }
                _experienceBand = value;
            }
        }
        
        public string DietStyleCode { get; set; } = string.Empty;
        public IReadOnlyCollection<TimeWindow> TimeWindows => _timeWindows.AsReadOnly();
        public IReadOnlyCollection<string> AllowedGear => _allowedGear;

        public void UpdateGoal(GoalCode newGoal)
        {
            // Использование this для доступа к свойству
            var oldGoal = this.PrimaryGoal;
            this.PrimaryGoal = newGoal;
            LogGoalChange(oldGoal, newGoal);
        }

        public void SetTimeWindows(IEnumerable<TimeWindow> windows)
        {
            _timeWindows.Clear();
            _timeWindows.AddRange(windows);
        }

        public bool Prefers(string gear)
        {
            // Использование this для доступа к приватному полю
            return this._allowedGear.Contains(gear);
        }

        public void AddGear(string gear)
        {
            if (string.IsNullOrWhiteSpace(gear))
            {
                throw new ArgumentException("Gear cannot be null or empty", nameof(gear));
            }
            
            // Использование this
            this._allowedGear.Add(gear);
        }
        
        // Деструктор для уменьшения счетчика
        ~AthleteAccount()
        {
            _accountCount--;
        }

        private static void LogGoalChange(GoalCode oldGoal, GoalCode newGoal)
        {
            Console.WriteLine($"    [AthleteAccount] Goal changed from {(int)oldGoal} to {(int)newGoal}");
        }
    }
}



